packages/view-types/spatial/src/ScaleBarLayer.js

```diff
@@ -0,0 +1,186 @@
+  length: { type: 'number', value: 0.085, compare: true },
+};
+/**
+ * @typedef LayerProps
+ * @type {Object}
+ * @property {String} unit Physical unit size per pixel at full resolution.
+ * @property {Number} size Physical size of a pixel.
+ * @property {Object} viewState The current viewState for the desired view.  We cannot internally use this.context.viewport because it is one frame behind:
+ * https://github.com/visgl/deck.gl/issues/4504
+ * @property {Array=} boundingBox Boudning box of the view in which this should render.
+ * @property {string=} id Id from the parent layer.
+ * @property {number=} length Value from 0 to 1 representing the portion of the view to be used for the length part of the scale bar.
+ */
+
+/**
+ * @type {{ new(...props: LayerProps[]) }}
+ * @ignore
+ */
+const ScaleBarLayer = class extends CompositeLayer {
+  renderLayers() {
+    const { id, unit, size, position, viewState, length, scaleFactor } = this.props;
+    const boundingBox = makeBoundingBox(viewState);
+    const { zoom } = viewState;
+    const viewLength = boundingBox[2][0] - boundingBox[0][0];
+    const barLength = viewLength * 0.05;
+
+    const adjustment = barLength * scaleFactor;
+
+    // This is a good heuristic for stopping the bar tick marks from getting too small
+    // and/or the text squishing up into the bar.
+    const barHeight = Math.max(
+      2 ** (-zoom + 1.5),
+      (boundingBox[2][1] - boundingBox[0][1]) * 0.007,
+    );
+
+    const numUnits = barLength * size;
+    const [yCoord, xLeftCoord] = getPosition(boundingBox, position, length);
+
+    const lengthBar = new LineLayer({
+      id: `scale-bar-length-${id}`,
+      coordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
+      data: [
+        [
+          [xLeftCoord - adjustment, yCoord],
+          [xLeftCoord + barLength, yCoord],
+        ],
+      ],
+      getSourcePosition: d => d[0],
+      getTargetPosition: d => d[1],
+      getWidth: 2,
+      getColor: [220, 220, 220],
+    });
+    const tickBoundsLeft = new LineLayer({
+      id: `scale-bar-height-left-${id}`,
+      coordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
+      data: [
+        [
+          [xLeftCoord - adjustment, yCoord - barHeight],
+          [xLeftCoord - adjustment, yCoord + barHeight],
+        ],
+      ],
+
+      getSourcePosition: d => d[0],
+      getTargetPosition: d => d[1],
+      getWidth: 2,
+      getColor: [220, 220, 220],
+    });
+    const tickBoundsRight = new LineLayer({
+      id: `scale-bar-height-right-${id}`,
+      coordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
+      data: [
+        [
+          [xLeftCoord + barLength, yCoord - barHeight],
+          [xLeftCoord + barLength, yCoord + barHeight],
+        ],
+      ],
+      getSourcePosition: d => d[0],
+      getTargetPosition: d => d[1],
+      getWidth: 2,
+      getColor: [220, 220, 220],
+    });
+
+    function snapToRound(value) {
+      const magnitude = Math.pow(10, Math.floor(Math.log10(value)));
```

Comment:
Can all or most of the math (or at least any new math added as part of this PR) be extracted to functions that can be unit-tested? It could also use comments to understand the intent.

Reply:
yes of course. All of this will not be in Vitessce -- it will be added to the viv codebase.